flink-optimizer/src/main/java/org/apache/flink/optimizer/operators/ReduceProperties.java

```diff
@@ -59,37 +65,69 @@ public DriverStrategy getStrategy() {
 	@Override
 	public SingleInputPlanNode instantiate(Channel in, SingleInputNode node) {
 		if (in.getShipStrategy() == ShipStrategyType.FORWARD ||
-				(node.getBroadcastConnections() != null && !node.getBroadcastConnections().isEmpty()))
+			(node.getBroadcastConnections() != null && !node.getBroadcastConnections().isEmpty()))
 		{
+			// adjust a sort (changes grouping, so it must be for this driver to combining sort
+			if(in.getSource().getOptimizerNode() instanceof PartitionNode) {
```

Comment:
Please refactor the code of this branch into a method called `injectCombinerBeforePartitioner()`. Some of the comments in the other class apply here as well.